File: extension/generate_icons.py
#!/usr/bin/env python3
"""Generate GramAI extension icons at all required sizes."""

import struct, zlib, os

def create_png(size, output_path):
    """Create a gradient purple icon PNG without PIL."""
    pixels = []
    cx, cy = size / 2, size / 2
    r = size / 2

    for y in range(size):
        row = []
        for x in range(size):
            # Circle mask
            dist = ((x - cx)**2 + (y - cy)**2) ** 0.5
            if dist > r - 0.5:
                row.extend([0, 0, 0, 0])  # transparent
                continue

            # Gradient: top-left #6366f1, bottom-right #8b5cf6
            t = (x + y) / (size * 2)
            pr = int(0x63 + (0x8b - 0x63) * t)
            pg = int(0x66 + (0x5c - 0x66) * t)
            pb = int(0xf1 + (0xf6 - 0xf1) * t)

            # Anti-alias edge
            alpha = int(min(1.0, (r - dist)) * 255)
            row.extend([pr, pg, pb, alpha])
        pixels.append(row)

    # Draw sparkle "✨" as simple white shapes
    def set_pixel(px, py, a=255):
        if 0 <= px < size and 0 <= py < size:
            pixels[py][px*4+3] = min(255, pixels[py][px*4+3])
            pixels[py][px*4]   = 255
            pixels[py][px*4+1] = 255
            pixels[py][px*4+2] = 255
            pixels[py][px*4+3] = a

    # Draw a simple "G" letter or star depending on size
    if size >= 32:
        # Draw a white star/cross in center
        m = size // 2
        arm = max(2, size // 8)
        for i in range(-arm, arm+1):
            for thickness in range(-max(1, size//16), max(1, size//16)+1):
                set_pixel(m + i, m + thickness)  # horizontal
                set_pixel(m + thickness, m + i)  # vertical
        # Diagonal arms (shorter)
        short = max(1, arm // 2)
        for i in range(-short, short+1):
            set_pixel(m + i, m + i)
            set_pixel(m + i, m - i)

    # Encode as PNG
    def png_chunk(name, data):
        c = zlib.crc32(name + data) & 0xffffffff
        return struct.pack('>I', len(data)) + name + data + struct.pack('>I', c)

    raw = b''
    for row in pixels:
        raw += b'\x00' + bytes(row)

    compressed = zlib.compress(raw, 9)
    ihdr = struct.pack('>IIBBBBB', size, size, 8, 6, 0, 0, 0)
    png = (
        b'\x89PNG\r\n\x1a\n' +
        png_chunk(b'IHDR', ihdr) +
        png_chunk(b'IDAT', compressed) +
        png_chunk(b'IEND', b'')
    )
    with open(output_path, 'wb') as f:
        f.write(png)
    print(f'  Created {output_path} ({size}x{size})')

def main():
    icons_dir = os.path.join(os.path.dirname(__file__), 'icons')
    os.makedirs(icons_dir, exist_ok=True)
    for size in [16, 32, 48, 128]:
        create_png(size, os.path.join(icons_dir, f'icon{size}.png'))
    print('All icons generated!')

if __name__ == '__main__':
    main()

File: extension/generate_icons.js
// Generate GramAI extension icons using Node.js (no dependencies)
const fs = require('fs');
const path = require('path');
const zlib = require('zlib');

function createPng(size, outputPath) {
  const pixels = [];
  const cx = size / 2;
  const cy = size / 2;
  const r = size / 2;

  for (let y = 0; y < size; y++) {
    const row = [];
    for (let x = 0; x < size; x++) {
      const dist = Math.sqrt((x - cx) ** 2 + (y - cy) ** 2);
      if (dist > r - 0.5) {
        row.push(0, 0, 0, 0);
        continue;
      }
      const t = (x + y) / (size * 2);
      const pr = Math.round(0x63 + (0x8b - 0x63) * t);
      const pg = Math.round(0x66 + (0x5c - 0x66) * t);
      const pb = Math.round(0xf1 + (0xf6 - 0xf1) * t);
      const alpha = Math.round(Math.min(1, r - dist) * 255);
      row.push(pr, pg, pb, alpha);
    }
    pixels.push(row);
  }

  if (size >= 32) {
    const m = Math.floor(size / 2);
    const arm = Math.max(2, Math.floor(size / 8));
    const setPixel = (px, py) => {
      if (px >= 0 && px < size && py >= 0 && py < size) {
        const i = px * 4;
        pixels[py][i] = 255;
        pixels[py][i + 1] = 255;
        pixels[py][i + 2] = 255;
        pixels[py][i + 3] = 255;
      }
    };
    for (let i = -arm; i <= arm; i++) {
      for (let t = -Math.max(1, Math.floor(size / 16)); t <= Math.max(1, Math.floor(size / 16)); t++) {
        setPixel(m + i, m + t);
        setPixel(m + t, m + i);
      }
    }
  }

  const crc32 = (buf) => {
    let c = 0xffffffff;
    for (let i = 0; i < buf.length; i++) {
      c ^= buf[i];
      for (let k = 0; k < 8; k++) c = (c >>> 1) ^ (0xedb88320 & -(c & 1));
    }
    return (c ^ 0xffffffff) >>> 0;
  };

  const chunk = (name, data) => {
    const len = Buffer.alloc(4);
    len.writeUInt32BE(data.length);
    const type = Buffer.from(name);
    const crc = Buffer.alloc(4);
    crc.writeUInt32BE(crc32(Buffer.concat([type, data])));
    return Buffer.concat([len, type, data, crc]);
  };

  let raw = Buffer.alloc(0);
  for (const row of pixels) {
    raw = Buffer.concat([raw, Buffer.from([0]), Buffer.from(row)]);
  }
  const compressed = zlib.deflateSync(raw, { level: 9 });
  const ihdr = Buffer.alloc(13);
  ihdr.writeUInt32BE(size, 0);
  ihdr.writeUInt32BE(size, 4);
  ihdr[8] = 8; ihdr[9] = 6; ihdr[10] = 0; ihdr[11] = 0; ihdr[12] = 0;

  const png = Buffer.concat([
    Buffer.from([0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a]),
    chunk('IHDR', ihdr),
    chunk('IDAT', compressed),
    chunk('IEND', Buffer.alloc(0)),
  ]);

  fs.writeFileSync(outputPath, png);
  console.log(`Created ${outputPath} (${size}x${size})`);
}

const iconsDir = path.join(__dirname, 'icons');
fs.mkdirSync(iconsDir, { recursive: true });
[16, 32, 48, 128].forEach(size => createPng(size, path.join(iconsDir, `icon${size}.png`)));
console.log('All icons generated!');
